baybe/simulation/lookup.py

```diff
@@ -0,0 +1,211 @@
+            If omitted, fake values will be used.
+
+            - pd.DataFrame: Each row contains the one parameter configuration and the
+              corresponding target values.
+            - callable: A callable that can be used to retrieve target values for given
+              parameter configurations. Expects a dataframe where each row corresponds
+              to one parameter configuration (columns names being the parameter names)
+              and must return a dataframe containing the corresponding target values
+              (column names being the target names), where each row holds the targets
+              for the corresponding parameter row.
+        impute_mode: Specifies how a missing lookup will be handled (only relevant
+            for dataframe lookups). See :func:`baybe.simulation.core.simulate_campaign`
+            for details.
+
+    Raises:
+        ValueError: If the provided lookup callable returns the wrong targets.
+    """
+    # If no lookup is provided, invent some fake values
+    if lookup is None:
+        add_fake_results(queries, targets)
+
+    # Compute the target values via a callable
+    elif isinstance(lookup, Callable):
+        _lookup_targets_from_callable(queries, [t.name for t in targets], lookup)
+
+    # Get the results via dataframe lookup (works only for exact matches)
+    # IMPROVE: Although its not too important for a simulation, this
+    #  could also be implemented for approximate matches
+    elif isinstance(lookup, pd.DataFrame):
+        _lookup_targets_from_dataframe(queries, targets, lookup, impute_mode)
+
+
+def _lookup_targets_from_callable(
+    queries: pd.DataFrame,
+    target_names: Collection[str],
+    lookup: _TCallableLookup,
+) -> None:
+    """Look up target values via a callable."""
+    # Evaluate the callable
+    responses = lookup(queries)
+
+    # Assert that all targets are contained in the response
+    if not (exp := set(target_names)).issubset(act := set(responses)):
+        raise ValueError(
+            f"The provided lookup callable yielded values for the labels {act} but "
+            f"required are values for {exp}."
+        )
+
+    # Insert the target values in-place
+    queries[responses.columns] = responses
+
+
+def _lookup_targets_from_dataframe(
+    queries: pd.DataFrame,
+    targets: List[Target],
+    lookup: _TCallableLookup,
+    impute_mode: Literal["error", "worst", "best", "mean", "random"] = "error",
+) -> None:
+    """Look up target values from a dataframe."""
+    target_names = [t.name for t in targets]
+    all_match_vals = []
+
+    # IMPROVE: speed up the matching using a join
+    for _, row in queries.iterrows():
+        ind = lookup[
+            (lookup.loc[:, row.index] == row).all(axis=1, skipna=False)
+        ].index.values
+
+        if len(ind) > 1:
+            # More than one instance of this parameter combination
+            # have been measured
+            warnings.warn(
```

Comment:
(quick info upfront: this all still is your original logic, no changes from my end)

I really like your idea 👍🏼 But I wonder what is good way to wrap this in and *intuitive + failsafe* interface. People should definitely be 100% aware that stuff is randomly sampled in case they simply forgot to deduplicate. On the other hand, having the possibility would be nice, as you said. One option might be via additional arguments. But that would bloat up our dense interfaces even further. Another would be to allow passing `Lookup` objects that are configurable. Any other/better ideas? In any case: since nothing has been changed, definitely stuff for a next PR (speaking of: refactoring is not yet over ;D)